core/lib/Cardano/Address/Errors.hs

```diff
@@ -0,0 +1,42 @@
+{-# LANGUAGE ExistentialQuantification #-}
+
+-- |
+-- Copyright: © 2020 IOHK
+-- License: Apache-2.0
+
+module Cardano.Address.Errors
+    ( -- $overview
+
+      -- * AddrError
+      AddrError(..)
+  ) where
+
+import Prelude
+
+import Cardano.Address
+    ( Address )
+
+import Control.Exception
+    ( Exception )
+
+-- $overview
+--
+-- This module provides error/exception types for Addresses.
+
+data AddrError = UnknownAddrType Address
+               | WrongInputSize Address
+                                Int -- the actual size
+               | PtrRetrieveError Address
+                                  String -- human readable error of underlying operation
+               | InvalidDerivationPath Address
+               | forall e . (Exception e, Show e) => DeserialiseError Address e
+
+instance Show AddrError where
+  show (UnknownAddrType addr)       = "UnknownAddrType "       ++ show addr
+  show (WrongInputSize addr int)    = "WrongInputSize "        ++ show addr ++ " " ++ show int
+  show (PtrRetrieveError addr str)  = "PtrRetrieveError "      ++ show addr ++ " " ++ str
+  show (InvalidDerivationPath addr) = "InvalidDerivationPath " ++ show addr
+  show (DeserialiseError addr e)    = "DeserialiseError "      ++ show addr ++ " " ++ show e
```

Comment:
It is generally **not** a good idea to manually define custom stock instances like `Show`, `Read` or `Ord`. Instead, consider privileging another type-class that is meant for displaying human-readable text input (Show really is made for debugging). 

More importantly, the malformed addresses should not be part of the output string. This can be an arbitrarily large / damaging payload which we certainly don't want to spit back to users. Instead, the error should gives examples of correct values whenever possible, or simply give hint about the failure in the input data.

Reply:
> It is generally not a good idea to manually define custom stock instances like

There is no way around it if there is an existential type inside

Reply:
What about using `StandaloneDeriving` and simply:

```hs
{-# LANGUAGE StandaloneDeriving #-}

deriving instance Show AddrError
```

with an adequate `Show` constraint on `e` ?

Reply:
Interesting, I didn't know about that trick. The problem is also existential types like these violate the haskell standard, so stock instances shouldn't work, afaiu.